### PR TITLE
[iOS] Migrate @UIApplicationMain attribute to @main

### PR DESCRIPTION
**Pull request by loic-sharma**
This migrates Flutter to use the `@main` attribute introduced in Swift 5.3. The `@UIApplicationMain` attribute is deprecated and will be removed in Swift 6. See: https://github.com/apple/swift-evolution/blob/main/proposals/0383-deprecate-uiapplicationmain-and-nsapplicationmain.md

This change is split into two commits:

1. https://github.com/flutter/flutter/pull/146707/commits/ad18797428ccbcfc14c6875118c44842b689ed6d - This updates the iOS app template and adds a migration to replace `@UIApplicationMain` uses with `@main`. 
2. https://github.com/flutter/flutter/pull/146707/commits/8ecbb2f29f419ed50fdebfbdcc069f3653638cc5 - I ran `flutter run` on each Flutter iOS app in this repository to verify the app migrates and launches successfully.

Part of https://github.com/flutter/flutter/issues/143044

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/ios/ios_project_migration_test.dart [hunk 1]
```diff
@@ -1408,6 +1409,104 @@ LD_RUNPATH_SEARCH_PATHS = $(inherited) /usr/lib/swift '@executable_path/../Frame
...
+    setUp(() {
+      memoryFileSystem = MemoryFileSystem();
+      testLogger = BufferLogger.test();
+      project = FakeIosProject();
+      appDelegateFile = memoryFileSystem.file('AppDelegate.swift');
+      project.appDelegateSwift = appDelegateFile;
+    });
+
+    testWithoutContext('skipped if files are missing', () async {
+      final UIApplicationmMainDeprecationMigration migration = UIApplicationmMainDeprecationMigration(
```
Comment (loic-sharma) [inline packages/flutter_tools/test/general.shard/ios/ios_project_migration_test.dart:1428]:
```suggestion
      final UIApplicationMainDeprecationMigration migration = UIApplicationMainDeprecationMigration(
```

File: dev/integration_tests/ios_host_app_swift/Host/AppDelegate.swift [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 import UIKit
 
-@UIApplicationMain
+@main
```
Comment (loic-sharma) [inline dev/integration_tests/ios_host_app_swift/Host/AppDelegate.swift:7]: This isn't a Flutter application. I updated this by hand.

File: examples/platform_channel_swift/ios/Runner/AppDelegate.swift [hunk 3]
```diff
@@ -19,7 +19,7 @@ enum MyFlutterErrorCode {
   static let unavailable = "UNAVAILABLE"
 }
 
-@UIApplicationMain
+@main
```
Comment (loic-sharma) [inline examples/platform_channel_swift/ios/Runner/AppDelegate.swift:22]: This file contains significant edits from the starter app template. The migration is flexible enough to be able to update this correctly.

File: packages/flutter_tools/lib/src/ios/migrations/uiapplicationmain_deprecation_migration.dart [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+class UIApplicationMainDeprecationMigration extends ProjectMigrator {
+  UIApplicationMainDeprecationMigration(
+    IosProject project,
+    super.logger,
+  ) : _appDelegateSwift = project.appDelegateSwift;
+
+  final File _appDelegateSwift;
+
+  @override
+  Future<bool> migrate() async {
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/ios/migrations/uiapplicationmain_deprecation_migration.dart:30]:
```suggestion
  Future<void> migrate() async {
```
I'm not sure why https://github.com/flutter/flutter/blob/8c12ba84705ba4a8056f83928fb83b251f194101/packages/flutter_tools/lib/src/ios/migrations/remove_bitcode_migration.dart#L19 has `bool`...

File: packages/flutter_tools/lib/src/ios/migrations/uiapplicationmain_deprecation_migration.dart [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    // Skip this migration if the project uses Objective-C.
+    if (!_appDelegateSwift.existsSync()) {
+      return true;
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/ios/migrations/uiapplicationmain_deprecation_migration.dart:33]: I think we usually `trace` here:
```suggestion
      logger.printTrace('AppDelegate.swift project not found, skipping @main migration.');
      return;
```

**Review by jmagman — APPROVED** [review]
Minor nits, LGTM!  Thank you for doing this!

**Comment by jmagman** [comment]
"Addresses https://github.com/flutter/flutter/issues/143044" if you make this "Fixes https://github.com/flutter/flutter/issues/143044" GitHub will link them and close the issue when this is merged.

File: packages/flutter_tools/lib/src/ios/migrations/uiapplicationmain_deprecation_migration.dart (hunk 4 above)
Comment (loic-sharma) [reply packages/flutter_tools/lib/src/ios/migrations/uiapplicationmain_deprecation_migration.dart]: Excellent catch! I fixed this and updated `RemoveBitcodeMigration` as well

File: packages/flutter_tools/lib/src/ios/migrations/uiapplicationmain_deprecation_migration.dart (hunk 5 above)
Comment (loic-sharma) [reply packages/flutter_tools/lib/src/ios/migrations/uiapplicationmain_deprecation_migration.dart]: Added

**Comment by vashworth** [comment]
> "Addresses #143044" if you make this "Fixes #143044" GitHub will link them and close the issue when this is merged.

@loic-sharma Did you happen to check if NSApplicationMain (the macOS version of this) was also deprecated?

I'm not sure if we want to mark the issue as fixed since it refers to both the iOS and macOS version. If we want to handle them separately, can you file a separate issue to address the macOS one?

File: packages/flutter_tools/lib/src/ios/migrations/uiapplicationmain_deprecation_migration.dart (hunk 4 above)
Comment (vashworth) [reply packages/flutter_tools/lib/src/ios/migrations/uiapplicationmain_deprecation_migration.dart]: Tests using `RemoveBitcodeMigration` need to updated to no longer expect bool

Example:
https://github.com/flutter/flutter/blob/839a6b3c8747557dd858b0237c482219573e2cea/packages/flutter_tools/test/general.shard/ios/ios_project_migration_test.dart#L906-L910

Which should fix the failing tests:
```
error • This expression has a type of 'void' so its value can't be used • packages/flutter_tools/test/general.shard/ios/ios_project_migration_test.dart:910:16 • use_of_void_result
   error • This expression has a type of 'void' so its value can't be used • packages/flutter_tools/test/general.shard/ios/ios_project_migration_test.dart:926:16 • use_of_void_result
   error • This expression has a type of 'void' so its value can't be used • packages/flutter_tools/test/general.shard/ios/ios_project_migration_test.dart:947:16 • use_of_void_result
   error • This expression has a type of 'void' so its value can't be used • packages/flutter_tools/test/general.shard/ios/ios_project_migration_test.dart:1432:14 • use_of_void_result
   error • This expression has a type of 'void' so its value can't be used • packages/flutter_tools/test/general.shard/ios/ios_project_migration_test.dart:1461:14 • use_of_void_result
   error • This expression has a type of 'void' so its value can't be used • packages/flutter_tools/test/general.shard/ios/ios_project_migration_test.dart:1490:14 • use_of_void_result
```

```
Failed to load "test/general.shard/ios/ios_project_migration_test.dart":
   test/general.shard/ios/ios_project_migration_test.dart:910:16: Error: This expression has type 'void' and can't be used.
           expect(await migration.migrate(), isTrue);
                  ^
   test/general.shard/ios/ios_project_migration_test.dart:926:16: Error: This expression has type 'void' and can't be used.
           expect(await migration.migrate(), isTrue);
                  ^
   test/general.shard/ios/ios_project_migration_test.dart:947:16: Error: This expression has type 'void' and can't be used.
           expect(await migration.migrate(), isTrue);
                  ^
   test/general.shard/ios/ios_project_migration_test.dart:1432:14: Error: This expression has type 'void' and can't be used.
         expect(await migration.migrate(), isTrue);
                ^
   test/general.shard/ios/ios_project_migration_test.dart:1461:14: Error: This expression has type 'void' and can't be used.
         expect(await migration.migrate(), isTrue);
                ^
   test/general.shard/ios/ios_project_migration_test.dart:1490:14: Error: This expression has type 'void' and can't be used.
         expect(await migration.migrate(), isTrue);
```

**Comment by hellohuanlin** [comment]
How do I trigger this migration for existing projects? Do I run a migration command? Or do I simply do `flutter clean` and then regenerate these files?

**Comment by jmagman** [comment]
> How do I trigger this migration for existing projects? Do I run a migration command? Or do I simply do `flutter clean` and then regenerate these files?

The migrations run when the app compiles, so every `flutter run` will do it.

File: packages/flutter_tools/lib/src/ios/migrations/uiapplicationmain_deprecation_migration.dart (hunk 4 above)
Comment (loic-sharma) [reply packages/flutter_tools/lib/src/ios/migrations/uiapplicationmain_deprecation_migration.dart]: Fixed! (... and ran the the tests locally this time 😅)

**Pull request merged**